cpp/comms/std/src/ucp_helper.h

```diff
@@ -53,165 +74,173 @@ static const int UCP_ANY_RANK = -1;
-                        ucp_tag_recv_info_t *info) {
+static void recv_callback(void *request, ucs_status_t status,
+                          ucp_tag_recv_info_t *info) {
   struct ucx_context *context = (struct ucx_context *)request;
   context->completed = 1;
 }
 
-void load_ucp_handle(struct comms_ucp_handle *ucp_handle) {
-  ucp_handle->ucp_handle =
-    dlopen("libucp.so", RTLD_LAZY | RTLD_NOLOAD | RTLD_NODELETE);
-  if (!ucp_handle->ucp_handle) {
-    ucp_handle->ucp_handle = dlopen("libucp.so", RTLD_LAZY | RTLD_NODELETE);
-    ASSERT(ucp_handle->ucp_handle, "Cannot open UCX library: %s\n", dlerror());
+/**
+ * Helper class for managing `dlopen` state and
+ * interacting with ucp.
+ */
+class comms_ucp_handler {
+ public:
+  comms_ucp_handler() {
+    load_ucp_handle();
+    load_send_func();
+    load_recv_func();
+    load_free_req_func();
+    load_print_info_func();
+    load_worker_progress_func();
   }
-  dlerror();
-}
 
-void close_ucp_handle(struct comms_ucp_handle *handle) {
-  dlclose(handle->ucp_handle);
-}
+  ~comms_ucp_handler() { dlclose(ucp_handle); }
 
-void assert_dlerror() {
-  char *error = dlerror();
-  ASSERT(error == NULL, "Error loading function symbol: %s\n", error);
-}
-
-void load_send_func(struct comms_ucp_handle *ucp_handle) {
-  ucp_handle->send_func = (ucs_status_ptr_t(*)(
-    ucp_ep_h, const void *, size_t, ucp_datatype_t, ucp_tag_t,
-    ucp_send_callback_t))dlsym(ucp_handle->ucp_handle, "ucp_tag_send_nb");
-  assert_dlerror();
-}
+ private:
+  void *ucp_handle;
 
-void load_free_req_func(struct comms_ucp_handle *ucp_handle) {
-  ucp_handle->req_free_func =
-    (void (*)(void *request))dlsym(ucp_handle->ucp_handle, "ucp_request_free");
-  assert_dlerror();
-}
+  dlsym_print_info print_info_func;
+  dlsym_rec_free req_free_func;
+  dlsym_worker_progress worker_progress_func;
+  dlsym_send send_func;
+  dlsym_recv recv_func;
+
+  void load_ucp_handle() {
+    ucp_handle = dlopen("libucp.so", RTLD_LAZY | RTLD_NOLOAD | RTLD_NODELETE);
+    if (!ucp_handle) {
+      ucp_handle = dlopen("libucp.so", RTLD_LAZY | RTLD_NODELETE);
+      ASSERT(ucp_handle, "Cannot open UCX library: %s\n", dlerror());
+    }
+    // Reset any potential error
+    dlerror();
+  }
 
-void load_print_info_func(struct comms_ucp_handle *ucp_handle) {
-  ucp_handle->print_info_func = (void (*)(ucp_ep_h, FILE *))dlsym(
-    ucp_handle->ucp_handle, "ucp_ep_print_info");
-  assert_dlerror();
-}
+  void assert_dlerror() {
+    char *error = dlerror();
+    ASSERT(error == NULL, "Error loading function symbol: %s\n", error);
+  }
 
-void load_worker_progress_func(struct comms_ucp_handle *ucp_handle) {
-  ucp_handle->worker_progress_func =
-    (int (*)(ucp_worker_h))dlsym(ucp_handle->ucp_handle, "ucp_worker_progress");
-  assert_dlerror();
-}
+  void load_send_func() {
+    send_func = (ucs_status_ptr_t(*)(
+      ucp_ep_h, const void *, size_t, ucp_datatype_t, ucp_tag_t,
+      ucp_send_callback_t))dlsym(ucp_handle, "ucp_tag_send_nb");
```

Comment:
another minor nitpick: we could use the typedef'd names here as well, right? (same applies to other `load_*` methods too)